app/models/miq_smart_proxy_worker/runner.rb

```diff
@@ -1,3 +1,69 @@
 class MiqSmartProxyWorker::Runner < MiqQueueWorkerBase::Runner
   self.delay_startup_for_vim_broker = true # NOTE: For smartproxy role
+
+  def do_before_work_loop
+    @tid = start_heartbeat_thread
+  end
+
+  def before_exit(message, _exit_code)
+    @exit_requested = true
+    #
+    # Stop the Heartbeat Thread
+    #
+    safe_log("#{message} Stopping Heartbeat Thread.")
+
+    #
+    # Wait for the Heartbeat Thread to stop
+    #
+    unless @tid.nil?
+      safe_log("#{message} Waiting for Heartbeat Thread to Stop.")
+      @tid.join(worker_settings[:heartbeat_thread_shutdown_timeout]) rescue nil
+    end
+  end
+
+  def start_heartbeat_thread
+    @exit_requested    = false
+    @heartbeat_started = Concurrent::Event.new
+    _log.info("#{log_prefix} Starting Heartbeat Thread")
+
+    tid = Thread.new do
+      begin
+        heartbeat_thread
+      rescue => err
+        _log.error("#{log_prefix} Heartbeat Thread aborted because [#{err.message}]")
+        _log.log_backtrace(err)
+        Thread.exit
+      ensure
+        @heartbeat_started.set
+      end
+    end
+
+    @heartbeat_started.wait
+    _log.info("#{log_prefix} Started Heartbeat Thread")
+
+    tid
+  end
+
+  def heartbeat_thread
+    @heartbeat_started.set
+    until @exit_requested do
+      heartbeat
```

Comment:
Should this also check to see that the main thread is still alive?

Reply:
@roliveri that would be beneficial.  I'd have to think about how to do that.  Note that this code was mostly adapted from the Event Monitor Thread in the EventCatcher runner.  I don't see that thread checking for life of the main thread either - should it?

Reply:
@jerryk55 Currently, the work thread is the main thread, and it spawns the heartbeat thread. So, the heartbeat thread may die if the worker thread terminates, but I'm not sure. If the heartbeat thread continues to run when the worker thread dies, then the heartbeat thread should check the worker thread. You probably have to do some testing to determine the behavior.

Reply:
@roliveri docs state that the peer threads exit if the main thread does but just to be sure I tested the theory by adding a Thread.exit to the main thread after the Heartbeat thread was started and the entire process shut down.  In my opinion the heartbeat thread does not need to check if the main thread is alive.  I think this is ready to be merged unless you have any other questions or issues.  Thanks.